lib/Reader/readerir.cpp

```diff
@@ -7666,97 +7706,186 @@ Type *GenIR::Vector4Ty;
+    }
+  } else {
+    return 0;
+  }
 }
 
 IRNode *GenIR::vectorEqual(IRNode *Vector1, IRNode *Vector2) {
-  assert(((Value *)Vector1)
-             ->getType()
-             ->getVectorElementType()
-             ->isFloatingPointTy());
-  assert(((Value *)Vector2)->getType() == ((Value *)Vector1)->getType());
-  return (IRNode *)LLVMBuilder->CreateFCmpOEQ(Vector1, Vector2);
+  assert(Vector2->getType() == Vector1->getType());
+  if (Vector1->getType()->getVectorElementType()->isFloatingPointTy()) {
+    return (IRNode *)LLVMBuilder->CreateFCmpOEQ(Vector1, Vector2);
+  }
+  if (Vector1->getType()->getVectorElementType()->isIntegerTy()) {
+    return (IRNode *)LLVMBuilder->CreateICmpEQ(Vector1, Vector2);
+  }
+  return 0;
 }
 
 IRNode *GenIR::vectorNotEqual(IRNode *Vector1, IRNode *Vector2) {
-  assert(((Value *)Vector1)
-             ->getType()
-             ->getVectorElementType()
-             ->isFloatingPointTy());
-  assert(((Value *)Vector2)->getType() == ((Value *)Vector1)->getType());
-  return (IRNode *)LLVMBuilder->CreateFCmpONE(Vector1, Vector2);
+  assert(Vector2->getType() == Vector1->getType());
+  if (Vector1->getType()->getVectorElementType()->isFloatingPointTy()) {
+    return (IRNode *)LLVMBuilder->CreateFCmpONE(Vector1, Vector2);
+  }
+  if (Vector1->getType()->getVectorElementType()->isIntegerTy()) {
+    return (IRNode *)LLVMBuilder->CreateICmpNE(Vector1, Vector2);
+  }
+  return 0;
+}
+
+IRNode *GenIR::vectorMax(IRNode *Vector1, IRNode *Vector2) {
+  assert(Vector2->getType() == Vector1->getType());
+  if (Vector1->getType()->getVectorElementType()->isFloatingPointTy()) {
+    std::vector<Type *> Types;
+    Types.push_back(Vector1->getType());
+    llvm::Function *Func = 0;
+    Func = Intrinsic::getDeclaration(JitContext->CurrentModule,
+                                     Intrinsic::maxnum, Types);
+    std::vector<Value *> Args;
+    Args.push_back(Vector1);
+    Args.push_back(Vector2);
+    return (IRNode *)LLVMBuilder->CreateCall(Func, Args);
+  }
+  return 0;
+}
+
+IRNode *GenIR::vectorMin(IRNode *Vector1, IRNode *Vector2) {
+  assert(Vector2->getType() == Vector1->getType());
+  if (Vector1->getType()->getVectorElementType()->isFloatingPointTy()) {
+    std::vector<Type *> Types;
+    Types.push_back(Vector1->getType());
+    llvm::Function *Func = 0;
+    Func = Intrinsic::getDeclaration(JitContext->CurrentModule,
+                                     Intrinsic::minnum, Types);
+    std::vector<Value *> Args;
+    Args.push_back(Vector1);
+    Args.push_back(Vector2);
+    return (IRNode *)LLVMBuilder->CreateCall(Func, Args);
+  }
+  return 0;
+}
+
+IRNode *GenIR::vectorBitOr(IRNode *Vector1, IRNode *Vector2) {
+  assert(Vector2->getType() == Vector1->getType());
+  if (Vector1->getType()->getVectorElementType()->isFloatingPointTy()) {
+    Type *ResultType = Vector1->getType();
+    Vector1 = (IRNode *)LLVMBuilder->CreateBitCast(Vector1, VectorIntTy);
```

Comment:
Okay. My understanding was that we need both the number of elements in the vector and the width in bits of those elements to match for this to be correct. It doesn't look like that's generally the case here: both the element type of `VectorIntTy` and the number of elements are hard-coded.